polaris-core/src/main/java/org/apache/polaris/core/catalog/PageToken.java

```diff
@@ -0,0 +1,145 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.catalog;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Represents a page token that can be used by operations like `listTables`. Clients that specify a
+ * `pageSize` (or a `pageToken`) may receive a `next-page-token` in the response, the content of
+ * which is a serialized PageToken.
+ *
+ * <p>By providing that in the next query's `pageToken`, the client can resume listing where they
+ * left off. If the client provides a `pageToken` or `pageSize` but `next-page-token` is null in the
+ * response, that means there is no more data to read.
+ */
+public class PageToken {
+
+  public final int offset;
+  public final int pageSize;
+
+  private static final String TOKEN_PREFIX = "polaris";
+  private static final int TOKEN_START = 0;
+  private static final int DEFAULT_PAGE_SIZE = 1000;
+
+  public static PageToken DONE = null;
+
+  public PageToken(int offset, int pageSize) {
+    this.offset = offset;
+    this.pageSize = pageSize;
+  }
+
+  /** Construct a PageToken from a plain limit */
+  public static PageToken fromLimit(int limit) {
+    return new PageToken(TOKEN_START, limit);
+  }
+
+  /** Construct a PageToken to read everything */
+  public static PageToken readEverything() {
+    return new PageToken(TOKEN_START, Integer.MAX_VALUE);
+  }
+
+  /** Deserialize a token string into a PageToken object */
+  public static PageToken fromString(String tokenString) {
+    if (tokenString == null) {
+      return PageToken.readEverything();
+    } else if (tokenString.isEmpty()) {
+      return PageToken.fromLimit(DEFAULT_PAGE_SIZE);
+    } else {
+      try {
+        String decoded =
+            new String(Base64.getDecoder().decode(tokenString), StandardCharsets.UTF_8);
+        String[] parts = decoded.split(":");
+
+        if (parts.length != 4 || !parts[0].equals(TOKEN_PREFIX)) {
+          throw new IllegalArgumentException("Invalid token format in token: " + tokenString);
+        }
+
+        int offset = Integer.parseInt(parts[1]);
+        int pageSize = Integer.parseInt(parts[2]);
+        int checksum = Integer.parseInt(parts[3]);
+        PageToken token = new PageToken(offset, pageSize);
+
+        if (token.hashCode() != checksum) {
+          throw new IllegalArgumentException("Invalid checksum for token: " + tokenString);
```

Comment:
Would it be good to provided checksum and expected checksum for debugging purpose?

Reply:
Also we are catching this exception and providing the token before rethrowing it. Would this unnecessarily duplicate?

Reply:
The spec specifically says tokens are `opaque`, so I don't know how much detail we necessarily need to provide here. The checksum not matching the content of the token could mean that the checksum was malformed, or the content was, or something else entirely. The important thing is that the token is no longer valid and must be rejected.

Reply:
Agreed. Assuming the token is always generated by Polaris, it should never be the case, unless the client is sending an entirely different value (by mistake). Since we are sending the token, we should be fine. Just curious, we are not logging anything in this class. Is it logged anywhere else as INFO or WARN?

Reply:
Tokens would get logged in the request, and returned in the response. Any logging here could be pretty noisy (potentially 2 tokens is constructed for each request/response), so I avoided it for now. Plus, there shouldn't be must logic in here. But I agree the parsing is pretty gnarly, so do you think we should add some logging?

Reply:
Assuming it is logged somewhere, then it should be fine. Essentially we should be able to identify rogue clients.